### DISCUSSION TITLE
A few questions about using the API

### DISCUSSION THREAD

**Discussion by dayunbao** [Mikubill/sd-webui-controlnet#1839] [Q&A] [answered]
I'm almost done switching over from Stability AI's API to Automatic1111/SD WebuiControlNet's API, and I have a few questions.

1. Do we need to pass `enabled` every time we call one of the `/sdapi/v1/` endpoints if we plan on using a ControlNet?  [This](https://github.com/Mikubill/sd-webui-controlnet/blob/efda6ddfd82ebafc6e1150fbb7e1f27163482a82/example/txt2img_example/api_txt2img.py) example has it, but other examples like [this](https://github.com/Mikubill/sd-webui-controlnet/wiki/API#integrating-sdapiv12img) one don't.
2. I've enabled multi ControlNet in the settings, but am not seeing how to add more ControlNets in the client.  I don't have any extra tabs appearing, and I don't see any way to add them.  I'm guessing that the API equivalent of multi ControlNets is by having multiple items in the `args` list?  Like this:
```
    "alwayson_scripts": {
        "controlnet": {
          "args": [
            {
              control_net_1
            },
            {
              control_net_2
            },
           {
              control_net_3
            },
          ]
        }
      }
```
And the number of items in `args` can't exceed the value of `Multi ControlNet: Max models amount`?
3. I'm still a little unsure about the pre-processor resolution and thresholds.  It seems like the thresholds mean different things to different models, when they are used at all.  Should I just leave them at their defaults?  In what situation would I need to send something other than the default values? 
4. I noticed that enabling pixel perfect in the client automatically sets the pre-processor resolution, but the `/controlnet/detect` endpoint doesn't accept `pixel_perfect`.  Is there any plan to add that?  I was thinking I could just always send `pixel_perfect=True` and not worry about setting the resolution if calling `/controlnet/detect`.  However, since the endpoint doesn't accept that parameter, I guess I have the same question as I do for thresholds.  When to change the default value?

**Comment by dayunbao** [comment]
Not sure if I did something wrong, but I feel like something isn't working correctly.  Using the API from the app I'm building, I was able to successfully use a ControlNet pre-processor directly (openpose), and then used the returned image as input for a text-to-image generation.  However, the returned images didn't seem related to the OpenPose input at all.  See below.

Openpose: 
![openpose_success](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/43b6df34-ff2c-45d6-9f51-2caee3858561)

Sample response images:
![response](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/476ca3b4-686d-4430-a55d-64dfa17998e4)


Request body:
`{

    "prompt": "Samurai cat standing in a post-apocalyptic wasteland.",
    "negative_prompt": "",
    "cfg_scale": 7,
    "sampler_name": "Euler",
    "batch_size": 6,
    "seed": 0,
    "steps": 100,
    "width": 512,
    "height": 512,
    "alwayson_scripts": {
        "controlnet": {
            "args": [
                {
                    "input_image": very_long_string,
                    "enabled": true,
                    "lowvram": false,
                    "pixel_perfect": true,
                    "module": "openpose",
                    "model": "control_v11p_sd15_openpose [cab727d4]",
                    "weight": 1,
                    "guidance_start": 0,
                    "guidance_end": 1,
                    "control_mode": 0,
                    "resize_mode": 1
                }
            ]
        }
    }
}`

Did I do something wrong?

**Reply by huchenlei** [reply]
I think the script name `ControlNet` is case sensitive. See here:
https://github.com/Mikubill/sd-webui-controlnet/blob/efda6ddfd82ebafc6e1150fbb7e1f27163482a82/scripts/controlnet.py#L220C1-L221C28

**Reply by dayunbao** [reply]
OK.  I'll try that.  All of the example requests in the repo use `controlnet`, though.

Thanks!

**Reply by huchenlei** [reply, marked as answer]
Sorry about that, I was just suspecting. Another thing I noticed is that if you are passing an openpose skeleton to ControlNet, you should select preprocessor `none`, intead of `openpose`.

**Reply by dayunbao** [reply]
I'll give that a shot.  Thanks again!

**Reply by taii241** [reply]
Sorry, I would like to ask you how to implement the change of predictor to none. Do you mean changing model: "openpose" to model: "none"?

**Comment by dayunbao** [comment]
> Sorry about that, I was just suspecting. Another thing I noticed is that if you are passing an openpose skeleton to ControlNet, you should select preprocessor `none`, intead of `openpose`.

That was the problem.  Thanks!
